### ISSUE TITLE
question: does this package support multiple server instances monitoring the same collection/file system?

### ISSUE THREAD

**Issue by lindleycb** [VeliovGroup/Meteor-Files#510]
question: does this package support multiple server instances monitoring the same collection/file system?  I've just tried it but I'm getting a worker error on the second instance.  I essentially have a backoffice app and a user app and I want them both to be able to see the same document store and both be able to upload new documents (with the backoffice app having more power/rights).  Is this going to be possible or should I look at other packages?  I already tried cfs which works but seems buggy around this area.

**Comment by dr-dimitru** [comment]
> does this package support multiple server instances monitoring the same collection/file system?

Yes, but storage should be either synchronized across servers or single separate storage should be used, like `GridFS`, `AWS:S3`, `Google Cloud`, or any other.

To be able to upload the file, "sticky sessions" (*a.k.a. session persistence*) must be used, same required for Meteor's DDP connection.

Start with reading the docs about our [Heroku-ready demo](https://github.com/VeliovGroup/Meteor-Files-Demos/tree/master/demo), Heroku by default has no persistent FS, and uses multi-server structure.

> I've just tried it but I'm getting a worker error on the second instance.

What error?

>  I essentially have a backoffice app and a user app and I want them both to be able to see the same document store and both be able to upload new documents (with the backoffice app having more power/rights). Is this going to be possible or should I look at other packages?

You should be good.

> I already tried cfs which works but seems buggy around this area.

`cfs` is deprecated


Hope this will help you. Let me know wdyt?

**Comment by lindleycb** [comment]
Sounds promising.  I'll have a proper look in the morning at the links you sent.  Both my apps will run on the same dedicated server so if I could stick to using the file system that would be best.  The files are relatively sensitive so the end client doesn't want to use Amazon, Google etc.  GridFS would probably be acceptable.  I've attached a screenshot of the error - but it doesn't seem to give much information as to the cause.  The backoffice and user app share the same database.  Files uploaded with the former (first to start) work, but with the later give this error.  My guess was this was down to a race condition between observers on the database?

<img width="1680" alt="screen shot 2017-10-19 at 22 59 51" src="https://user-images.githubusercontent.com/5908831/31796711-c13e757e-b522-11e7-9cea-27258cc8b81a.png">

**Comment by lindleycb** [comment]
P.S. Thanks for responding so quickly!

**Comment by dr-dimitru** [comment]
I believe the SCP header doesn't allow to create `blob:` schema required by WebWorker used in a package.
You can update SCP header to allow `blob:*` or disable WebWorker with [`allowWebWorkers: false`](https://github.com/VeliovGroup/Meteor-Files/wiki/Insert-(Upload)).

**Comment by lindleycb** [comment]
I added allowWebWorkers: false to the insert and the error went away, the database got the document entry and I was just about to jump for joy when I realised the file hadn't appeared in the uploads folder.  Am I missing something?

**Comment by dr-dimitru** [comment]
1. What is your uploads folder?
 2. Show me *Server* logs

**Comment by lindleycb** [comment]
This is still running locally so the uploads folder is just on my desktop.  Here come the logs:

client:
<img width="1677" alt="screen shot 2017-10-20 at 08 31 22" src="https://user-images.githubusercontent.com/5908831/31809820-4818d51e-b571-11e7-9dd4-53803909e13d.png">

server on the user app where I uploaded it:
![screen shot 2017-10-20 at 08 31 30](https://user-images.githubusercontent.com/5908831/31809846-63631550-b571-11e7-8c8b-9926b5723ae7.png)

server on the backoffice app (whose local database it is:
![screen shot 2017-10-20 at 08 31 39](https://user-images.githubusercontent.com/5908831/31809868-757860d8-b571-11e7-9c41-b0edfd868e78.png)

**Comment by lindleycb** [comment]
Oddly that one did get created in the folder but is corrupt - this is exactly the behaviour I've been seeing with cfs.  The first chunk or so of the pdf seems to be \0s

**Comment by lindleycb** [comment]
Sorry forgot to attach the screenshot:
![screen shot 2017-10-20 at 08 36 31](https://user-images.githubusercontent.com/5908831/31810011-107f184c-b572-11e7-810c-b2dbbc419371.png)

**Comment by lindleycb** [comment]
Just tried again with a smaller image file.  Similar trace but no file at all.

<img width="1678" alt="screen shot 2017-10-20 at 08 40 13" src="https://user-images.githubusercontent.com/5908831/31810096-6a8795c6-b572-11e7-86b4-a4c8462e6b0b.png">

![screen shot 2017-10-20 at 08 40 23](https://user-images.githubusercontent.com/5908831/31810109-731ab060-b572-11e7-98cd-48330c9a22be.png)

![screen shot 2017-10-20 at 08 40 30](https://user-images.githubusercontent.com/5908831/31810120-787eb196-b572-11e7-890d-4ed6bcb4fc4c.png)

**Comment by dr-dimitru** [comment]
On dev and prod use absolute path for the storage, from the [FAQ](https://github.com/VeliovGroup/Meteor-Files#faq):

>  __Note: All files will be removed as soon as your application rebuilds or you run meteor reset.__ To keep your storage persistent during development use an absolute path outside of your project folder, e.g. `/data` directory.

Also see:
 - https://github.com/VeliovGroup/Meteor-Files/issues/250

**Comment by lindleycb** [comment]
Hi, sorry if I didn't make this clear.  On dev the uploads folder is a folder on my Desktop (not in either of the apps local areas) and in prod its a folder in /opt (also not within the apps area - so presumably not housekept by Meteor).

**Comment by lindleycb** [comment]
The zero byte thing is pretty much exactly what I was seeing in cfs - seemed to particularly effect zips and pdf and from what I could see the chunks were uploading OK and the issue appeared to be around copying the chunks to reassemble the file - perhaps the 2 servers were competing to do this.

**Comment by dr-dimitru** [comment]
> The zero byte thing is pretty much exactly what I was seeing in cfs - seemed to particularly effect zips and pdf and from what I could see the chunks were uploading OK and the issue appeared to be around copying the chunks to reassemble the file - perhaps the 2 servers were competing to do this.

Try to switch between transports (*not sure if it's related, but worth to try*), by default DDP is used, try HTTP, see [`settings.transport `](https://github.com/VeliovGroup/Meteor-Files/wiki/Insert-(Upload))

**Comment by lindleycb** [comment]
Sadly no change regardless of whether transport is http or ddp.

I did notice this though on the user app log when uploading the small file (which doesn't appear in the folder):
![screen shot 2017-10-20 at 17 31 53](https://user-images.githubusercontent.com/5908831/31831830-bd6e93ee-b5bc-11e7-959f-7903db555a40.png)

Is the "removing unfinished upload" significant?
The other (backoffice) server log looks like this:
![screen shot 2017-10-20 at 17 32 16](https://user-images.githubusercontent.com/5908831/31831859-d56ffcb2-b5bc-11e7-94b8-36443f1784a2.png)

**Comment by dr-dimitru** [comment]
> Sadly no change regardless of whether transport is http or ddp.

 1. Have you tried other files?
 2. Does hashsum of the file before and after upload different? (*this is one of our main tests*)
 3. What version of Meteor and Meteor-Files you're on?

> Is the "removing unfinished upload" significant?

Not okay, means upload was interrupted by something or exceeded upload TTL

**Comment by lindleycb** [comment]
Scrap the problem with pdf file corruption - you were right I was being an idiot and uploading an already corrupted pdf - sorry!  md5s do now match on successful uploads.

However, I'm still seeing the file being deleted when uploaded by the user app rather than the backoffice app.  This time I left the finder open as the large pdf uploaded and watched it appear and disappear.  I suspect then that with small files this happens too fast for finder to be notified so it looks like the file never uploaded but actually I think it's being created by one app and deleted by the other.

**Comment by lindleycb** [comment]
Forgot to answer the question about versions: all latest and greatest stable i.e. Meteor 1.5.2.2, ostrio:files 1.9.0

**Comment by dr-dimitru** [comment]
> Scrap the problem with pdf file corruption - you were right I was being an idiot and uploading an already corrupted pdf - sorry!

It happens

> However, I'm still seeing the file being deleted when uploaded by the user app

Check logs, every move is logged, if you're seeing something like: `removing unfinished upload` it's called by *Client*, probably `.abort()` method somehow got triggered.

Is it app? Android? iOS? If it's a browser, what browser?

**Comment by lindleycb** [comment]
Yea you're right - idiots happen all the time :-)

Currently it's browser and I'm testing it from Firefox (developer) and Chrome and getting the same failures.

I am seeing removedUnfinishedUpload on the user app (where I'm uploading the file):

> I20171020-23:28:22.200(1)? [FilesCollection] [Upload] [finish(ing)Upload] -> /Users/chris/Desktop/rk_document_uploads2/6pMr3hcbKEmqP2Teb.pdf
> I20171020-23:28:22.202(1)? [FilesCollection] [Upload] [finish(ed)Upload] -> /Users/chris/Desktop/rk_document_uploads2/6pMr3hcbKEmqP2Teb.pdf
> I20171020-23:28:22.204(1)? [FilesCollection] [_preCollectionCursor.observe] [removed]: 6pMr3hcbKEmqP2Teb
> I20171020-23:28:22.207(1)? [FilesCollection] [_preCollectionCursor.observe] [removeUnfinishedUpload]: 6pMr3hcbKEmqP2Teb

on the other app I see:

> I20171020-23:28:22.202(1)? [FilesCollection] [_preCollectionCursor.observe] [changed]: 6pMr3hcbKEmqP2Teb
> I20171020-23:28:22.204(1)? [FilesCollection] [_preCollectionCursor.observe] [removed]: 6pMr3hcbKEmqP2Teb

Timestamps may be significant I guess.

**Comment by dr-dimitru** [comment]
You've posted Server's logs, what about Client?
Does it work here - [our testing files app](https://files.veliov.com)?

**Comment by dr-dimitru** [comment]
Just a reference (*to understand what is happening*) - https://github.com/VeliovGroup/Meteor-Files/blob/b5e1a6bf8560f47ab4361b0a2b2b88d55920264e/server.js#L315

**Comment by lindleycb** [comment]
Thanks for the reference.  I've tried a couple of things today but I'm still no wiser.

1. I tried to reproduce the problem using a minimal example. `demo-simplest-upload` was perfect for the job.  The only change I made was to set the storage path to somewhere outside of the app's folder.  I ran the first instance with `meteor run` and the second with `export MONGO_URL=mongodb://localhost:3001/meteor; meteor run --port 3005`.  No matter which files I uploaded though I couldn't get it to fail, even after to upgrading the demo to the latest meteor etc. so there must be something different about my situation.  In my case it's 2 different apps not 2 instances of the same but both use the same package to handle the documents so the collection declaration etc. are identical.  Meteor accounts are involved but from what I can see that shouldn't change anything.

1. Given the code reference above I debugged into the app and stopped it on this line on the user app (the one that fails).  I found that doc always contained just an id (no isFinished) and that if I manually set isFinished to true then the document wasn't removed and all was as it should be.

Have you any idea what might be the difference between the behavior in simple demo app and that in my apps?  Which are the settings that most effect the isFinished flag?

**Comment by dr-dimitru** [comment]
I'll start debugging where `isFinished` is set to `true` - https://github.com/VeliovGroup/Meteor-Files/blob/b5e1a6bf8560f47ab4361b0a2b2b88d55920264e/server.js#L946-L966

> Have you any idea what might be the difference between the behavior in simple demo app and that in my apps?

 1. The order of hooks is misplaced and `_finishUpload` called too late
 2. You're using packages deeply integrated into `Meteor.Collection`s, like db-rules, simple-schema, etc. I have't used it for a while, so I can not tell what may be broken
 3. As mentioned [here](https://github.com/VeliovGroup/Meteor-Files/issues/511#issuecomment-338393682) order of packages at `.meteor/packages` file - matters. Try to move `ostrio:files` to the top of the file.

**Comment by dr-dimitru** [comment]
Hello @lindleycb ,

Any news on your end?

**Comment by lindleycb** [comment]
I tried moving the package to the very top but was still getting the same issue.

I'm afraid, time pressures on this project have meant that I've had to abandon Meteor-Files for this use case and start using an alternative package that stores the files in GridFS using the databases locking mechanism.

I'll still be using Meteor-Files on other projects that don't have multiple server instances though and I am very grateful for your responsiveness and the support you've provided on this issue.  Much appreciated!

**Comment by dr-dimitru** [comment]
Hello @lindleycb ,

Thank you for update on this. Sorry to hear that, and will be happy to see you using it on the next proj.
If you wish a minimal reproduction rep is welcome.

For further investigation on our end we need:
 1. Have you tried `http` transport for upload? (*in a logs I see only* `ddp`)
 2. Contents of `.meteor/packages`
 3. Contents of `.meteor/versions`
 4. Contents of `.meteor/release`

Thank you.

**Comment by lindleycb** [comment]
Thanks.  I did try specifying `http` transport but no change and I'd hoped the simple example would provide a minimal reproduction but I couldn't get it to fail.

Our apps include loads of packages because they, particularly the backoffice, touch on so many areas.  It also has to still include cfs to handle the documents that are already on the system as well as the new ones so narrowing it down could be difficult.

If I get a chance tho I'll have another look at a minimal reproduction - I've got a lot on at the moment though so it could be a while!
